### ISSUE TITLE
Support HTTP trailers

### ISSUE THREAD

**Issue by mhils** [mitmproxy/mitmproxy#2474]
We currently do not support HTTP/1.x or HTTP/2.0 trailers. HTTP trailers are basically additional headers sent after the body. https://discourse.mitmproxy.org/t/http-2-trailers-not-supported/557/2 mentions that gRPC is doing this, so this seems like a reasonable feature we should support. I think the rough implementation plan looks as follows:

- Add `HTTPFlow.trailers` attribute (which is a `http.Headers` instance or None)
- Modify HTTP/2.0 layer to support trailers.
- Wait if anyone ever requests HTTP/1.0 trailer support.

**Comment by Kriechi** [comment]
hyper-h2 already provides this information as event:
https://python-hyper.org/h2/en/stable/api.html#h2.events.TrailersReceived

**Comment by Mengjianhua-c** [comment]
Dear developer, is there any plan to support HTTP trailers? I also encountered this problem when trying to intercept grpc, very much hope that the development team solves this problem.

**Comment by emaayan** [comment]
hi.. so does this mean i cannot use MITMProxy with GRPC?

**Comment by nzinfo** [comment]
I've been add an experimental trailers header support in my personal branch, see. https://github.com/coreseekdev/mitmproxy   branch t/coreseek , commit cc6ae764250bff92ea4f5ae0231a7c9b3c09659f & d57eb720af767441fdedc89c3d9d9b0dfa726106 . It was mixed with some notes, so make a pull request seems a bit difficult to me .

**Comment by mhils** [comment]
Very cool, thanks for linking this here @nzinfo! This looks like a good start. We'd need to extend this to also store trailers as part of a flow. If anyone wants to give this a try, I'd be happy to help with a PR.

**Comment by mhils** [comment]
As @Prinzhorn pointed out in #4211, the first steps are completed here, now it's mostly a matter of exposing it in the UI. 😃

**Comment by Kriechi** [comment]
This needs to be reevaluated due to the new sans-io protocol stack.

**Comment by mhils** [comment]
Yes - we've just merged our new proxy core, which currently does not support trailers yet. I'm sorry about the regression, but the benefits in other areas clearly outweigh this and we need to ship at some point. If anyone wants to give this a stab again, I'd be happy to help!
